### DISCUSSION TITLE
How to store `ws` for each connected client

### DISCUSSION THREAD

**Discussion by StarMonk** [uNetworking/uWebSockets.js#843] [Q&A] [answered]
I am trying to store data received from websocket to REDIS and maintain ws relationship between the user and WebSocket id

But the only option, I see is: `getRemoteAddressAsText`, but this option is unreliable, since multiple users can have the same address.

`ws` returns `uWS.WebSocket {}` and I looked into the WebSocket interface, but I can't find any unique Identifier.

**Comment by e3dio** [comment, marked as answer]
```javascript
let id = 0;
const wsMap = new Map();

app.ws('/', {
   open: ws => {
      ws.id = ++id;
      wsMap.set(id, ws);
      console.log('open', ws.id);
   },
   close: (ws, code) => {
      wsMap.delete(ws.id);
      console.log('close', ws.id, code);
   }  
});

wsMap.get(id).send('Hello');
```

**Reply by StarMonk** [reply]
I didn't realize, I can set one by my own :) 

Thanks a lot

**Reply by StarMonk** [reply]
Is there anyway, I can pass extra headers during `open` and read it from `open`? like `ws.req.headers` <- something like this

**Reply by StarMonk** [reply]
@e3dio - I tried looking into Upgrade example: https://github.com/uNetworking/uWebSockets.js/blob/master/examples/Upgrade.js

But can't determine, if I can set the extra value there which I can read from `open`.

**Reply by e3dio** [reply]
```javascript
// pass values from ws upgrade handler

app().ws('/', {
   upgrade: (res, req, context) => { // request was made to open websocket, res req have all properties for request, cookies etc
      // add code here to determine if ws request should be accepted or denied
      // deny request with "res.writeStatus('401').end()"
      res.upgrade( // upgrade to websocket
         { ip: res.getRemoteAddressAsText() }, // 1st argument sets which properties to pass to ws object, in this case ip address
         req.getHeader('sec-websocket-key'),
         req.getHeader('sec-websocket-protocol'),
         req.getHeader('sec-websocket-extensions'), // 3 headers are used to setup websocket
         context // also used to setup websocket
      )
   },
   open: ws => {
      console.log(ws.ip)
   }
});
```

**Reply by StarMonk** [reply]
Thanks a ton, @e3dio - I will try it out.
